daemon/emer-aggregate-tally.c

```diff
@@ -305,18 +317,25 @@ emer_aggregate_tally_constructed (GObject *object)
                            "aggregate-events.db",
                            NULL);
   if (!emer_aggregate_tally_init_db (self, path, &error))
-    g_error ("Failed to initialize %s: %s", path, error->message);
-}
+    {
+      g_warning ("Failed to initialize %s: %s; trying to delete and recreate it",
+                 path, error->message);
+      g_clear_error (&error);
 
-static gboolean
-close_db (sqlite3  *db,
-          GError  **error)
-{
-  if (db == NULL)
-    return TRUE;
+      close_db (self->db, NULL);
+      self->db = NULL;
 
-  return CHECK (sqlite3_close (db));
+      if (g_unlink (path) != 0)
+        g_warning ("Failed to unlink %s: %s", path, g_strerror (errno));
+
+      if (!emer_aggregate_tally_init_db (self, path, &error))
+        {
+          g_warning ("Failed to initialize %s again: %s",
+                     path, error->message);
+        }
 
+      /* ...and just soldier on. */
```

Comment:
I guess my concern is that if it crashed it will just crash in a loop, so it is nice to try to recover. The other idea I had was to switch to an in-memory database. But maybe it should indeed crash.

A third option would be to propagate the error out to the caller, who can decide whether to try again.

A downside of trashing the db on error might be that it's harder to grab a broken dB to inspect it. Perhaps I should drop this bit for now...

Reply:
I think the attempt to recover is good. What I'm saying is that if it tries to recover but still can't initialize the database, should it continue? What is the `EmerAggregateTally` object going to do in that case? Segfault somewhere else trying to access `self->db`? From the daemon's perspective I think there could be value in carrying on since there's more to it than aggregate metrics, but it seems like it would just crash somewhere eventually.

Reply:
Sorry, my comment should really be 2 lines higher in the `!emer_aggregate_tally_init_db` block. If it's actually recovered, it should carry on.

Reply:
Funny, some guy has written this exact logic before for the old storage format.

18442e17661afbf92e9c1f7555ed3410570106f9

and that time I opted to `g_error` after a second failure.

Reply:
I reworked this to `g_error` if deleting the db and trying again fails.

Reply:
I think that makes sense. `g_error` is not the best error handler, but at some point the program just has to give up. You could handle this in `emer_daemon_constructed` like you did before, but I think you'd have to change `EmerAggregateTally` to `GInitable`. Probably not worth the effort right now.

Reply:
Agreed.